activerecord/lib/active_record/connection_handling.rb

```diff
@@ -243,15 +243,22 @@ def connected_to?(role:, shard: ActiveRecord::Base.default_shard)
     # Clears the query cache for all connections associated with the current thread.
     def clear_query_caches_for_current_thread
       connection_handler.each_connection_pool do |pool|
-        pool.connection.clear_query_cache if pool.active_connection?
+        pool.connection.clear_query_cache
```

Comment:
I don't think this change belongs in this extracted PR. 

This causes an error when trying to operate on a freshly checked-out connection:

```ruby
ActiveRecord::ConnectionAdapters::AbstractAdapter.set_callback :checkout, :after, lambda { |conn|
  conn.exec_query("SET application_name = 'A named connection'", "Set application name")
}
```

...leads to recursively trying to checkout a new connection until the connection pool is exhausted.

Reply:
Could you provide a backtrace or repro?

Reply:
`exec_query` calls `clear_query_caches_for_current_thread`; with this change every pool will (permanently*) pin a connection the first time any query occurs, even if that connection is otherwise not in use.

I assume this applies more once the cache is thread/pool scoped, but not yet?

\* until the end of executor scope

Reply:
The query cache is already kept in the pool. This should probably just be `pool.clear_query_cache`.

Reply:
https://github.com/rails/rails/pull/51151

Reply:
> The query cache is already kept in the pool

🤦🏻‍♂️ I definitely knew that